lib/auth/touchid/api_darwin.go

```diff
@@ -0,0 +1,211 @@
+	defer func() {
+		C.free(unsafe.Pointer(req.label))
+		C.free(unsafe.Pointer(req.app_label))
+		C.free(unsafe.Pointer(req.app_tag))
+	}()
+
+	var errMsgC, pubKeyC *C.char
+	defer func() {
+		C.free(unsafe.Pointer(errMsgC))
+		C.free(unsafe.Pointer(pubKeyC))
+	}()
+
+	if res := C.Register(req, &pubKeyC, &errMsgC); res != 0 {
+		errMsg := C.GoString(errMsgC)
+		return nil, errors.New(errMsg)
+	}
+
+	pubKeyB64 := C.GoString(pubKeyC)
+	pubKeyRaw, err := base64.StdEncoding.DecodeString(pubKeyB64)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &CredentialInfo{
+		CredentialID: credentialID,
+		publicKeyRaw: pubKeyRaw,
+	}, nil
+}
+
+// rpID are domain names, so it's safe to assume they won't have spaces in them.
+// https://www.w3.org/TR/webauthn-2/#relying-party-identifier
+const labelSeparator = " "
+
+func makeLabel(rpID, user string) string {
+	return rpID + labelSeparator + user
+}
+
+func splitLabel(label string) (string, string) {
+	idx := strings.Index(label, labelSeparator)
+	if idx == -1 {
+		return "", ""
+	}
+	rpID := label[0:idx]
+	user := label[idx+1:]
+	return rpID, user
+}
+
+func (*touchIDImpl) Authenticate(credentialID string, digest []byte) ([]byte, error) {
+	var req C.AuthenticateRequest
+	req.app_label = C.CString(credentialID)
+	req.digest = C.CString(string(digest))
+	req.digest_len = C.int(len(digest))
+	defer func() {
+		C.free(unsafe.Pointer(req.app_label))
+		C.free(unsafe.Pointer(req.digest))
+	}()
+
+	var sigOutC, errMsgC *C.char
+	defer func() {
+		C.free(unsafe.Pointer(sigOutC))
+		C.free(unsafe.Pointer(errMsgC))
+	}()
+
+	if res := C.Authenticate(req, &sigOutC, &errMsgC); res != 0 {
+		errMsg := C.GoString(errMsgC)
+		return nil, errors.New(errMsg)
+	}
+
+	sigB64 := C.GoString(sigOutC)
+	return base64.StdEncoding.DecodeString(sigB64)
+}
+
+func (*touchIDImpl) FindCredentials(rpID, user string) ([]CredentialInfo, error) {
+	infos, res := findCredentialsImpl(rpID, user, func(filter C.LabelFilter, infosC **C.CredentialInfo) C.int {
+		return C.FindCredentials(filter, infosC)
+	})
+	if res < 0 {
+		return nil, fmt.Errorf("failed to find credentials: status %d", res)
+	}
+	return infos, nil
+}
+
+func findCredentialsImpl(rpID, user string, find func(C.LabelFilter, **C.CredentialInfo) C.int) ([]CredentialInfo, int) {
+	var filterC C.LabelFilter
+	if user == "" {
+		filterC.kind = C.LABEL_PREFIX
+	}
+	filterC.value = C.CString(makeLabel(rpID, user))
+	defer C.free(unsafe.Pointer(filterC.value))
+
+	var infosC *C.CredentialInfo
+	defer C.free(unsafe.Pointer(infosC))
+
+	res := find(filterC, &infosC)
+	if res < 0 {
+		return nil, int(res)
+	}
+
+	start := unsafe.Pointer(infosC)
+	size := unsafe.Sizeof(C.CredentialInfo{})
+	infos := make([]CredentialInfo, res)
+	for i := 0; i < int(res); i++ {
+		// IMPORTANT: The defer below is used to free the pointers inside infos.
+		// It relies on the fact that we never error out of the function after
+		// this point, otherwise some instances would leak.
```

Comment:
Why not just copy all the data from `infosC` first instead?

Reply:
That's essentially what we are doing here. I thought it better to avoid looping twice, given the somewhat unusual unsafe Pointer arithmetic involved.

Reply:
I think it would be less scary if we copied and freed the data in our loop; I also just realized that I don't know the perf characteristics of issuing a bunch of `defer`s, so doing it ourselves is at least a bit more obvious.

Reply:
Gotcha, I think I understood it better now. Changed the loop to read the data and free infos at the start, should look a bit safer now.